src/wallet/wallet2.cpp

```diff
@@ -1559,6 +1559,23 @@ void wallet2::fast_refresh(uint64_t stop_height, uint64_t &blocks_start_height,
   }
 }
 
+
+bool wallet2::add_address_book_row(const cryptonote::account_public_address &address, const crypto::hash &payment_id, const std::string &description)
+{
+  wallet2::address_book_row a;
+  a.m_address = address;
+  a.m_payment_id = payment_id;
+  a.m_description = description;
+  
+  int key = (m_address_book.empty())? 0 : m_address_book.rbegin()->first;
+  bool r =  m_address_book.emplace(++key,a).second;
+  return r;
```

Comment:
I don't really mind the map at all, it's just this weird code that doesn't strike me as safe. It might well be safe, but please explain how it is first (ie, if deleting a row not at the end, that can't "collapse" the rest, since the keys can't change. Therefore, the row isn't really a row (and maybe it's not really a row, more like a dummy key, in which case it'd be ok). Also, can the map's rbegin point to not the highest key (in which case the emplace might nuke an existing row). I'd be OK with that code if this was checked and commented as being correct. The deque thing is just a way to make it obviously correct.